Axial to planar gradiometer transformation

contsili opened the pull request:
#### Reference issue (if any)

Fixes #9609 .

#### What does this implement/fix?

-  Store the canonical CTF and Neuromag sensor definitions in a txt file
- Extent interpolate_to() to MEG sensors to accommodate axial to planar gradiometer transformation (and back)
- Since interpolate_to() works similarly to interpolate_bads() I suggest we create two new functions in interpolation.py: _interpolate_to_meeg that uses the 'MNE' interpolation method 



#### Additional information
The channel positions and orientations are adopted from fieldtrip: https://github.com/fieldtrip/fieldtrip/blob/master/template/gradiometer

I am not sure if for the interpolation we need the coil positions and orientations. For clarity: a gradiometer is ONE channel but has TWO coils. A magnetometer is ONE channel and has ONE coil.

As an expectation I have that I will plot ERPs in the ctf and neuromag format. Also I want to create topoplots like: https://www.fieldtriptoolbox.org/assets/img/tutorial/eventrelatedaveraging/figure8.png

larsoner commented:
Looks like you're making some progress, let me know when you'd like some feedback!

contsili commented:
Hi @larsoner! After a while, I finally found some time :)

I think this is a good moment for some feedback.

Issues / Questions I encountered:
**1. Handling ch_type when interpolating from CTF → Neuromag**

My initial understanding was that when we interpolate from CTF to Neuromag, the ch_type should already be known.

This is because, CTF systems have reference sensors that should not be interpolated and Neuromag systems have gradiometers and magnetometers, so we need to know the type of each channel.

That's why I added the ch_type parameter in the .txt montage files — to explicitly specify this.

**2. Scope of interpolation: only gradiometers?**
Following (1), I assumed we would only interpolate gradiometers (i.e., axial ↔ planar), and not magnetometers.

Based on this, I thought _map_eeg_and_meg_channels would use info_from and info_to objects containing only the gradiometer channels. This was another reason for adding the ch_type parameter in the montage.

**3. Limitations of make_dig_montage**
All the montages are created via make_dig_montage.

From the documentation of `make_dig_montage`:
```
 .. note::
            For custom montages without fiducials, this parameter must be set
            to ``'head'``.
```
 I understood that for my setup (no fiducials) I should use a custom montage.

However: running `montage = make_dig_montage(ch_pos=ch_pos, coord_frame="head")`  does not parse the ch_type or ori and  my channels are read as generic EEG labels, e.g., `standard_montage.ch_names[0] = 'EEG #1'` etc.

**4. Custom _meg() function and related problems**
Because of (3), I created a custom` _meg() `function to parse the .txt files.

But now I run into new problems. When I run `interpolate_to()` :


`_validate_type(sensors, DigMontage, "sensors")` leads to the error: `AttributeError: 'CustomMontage' object has no attribute 'get_positions'`
` ch_pos = sensors.get_positions().get("ch_pos", {})` leads to the error: `sensors must be an instance of DigMontage, got <class 'mne.channels._standard_montage_utils._meg.<locals>.CustomMontage'> instead`

**5. I added _meg() inside _standard_montage_utils but am I allowed to change such private function ?**

**6. I am trying to follow how interpolate_to() code style and structure is already built.**

However, this forces us to diverge a bit from the standard `interpolate_bads `recipe

One idea I had would be to encapsulate some logic in a new function `interpolate_to_meeg()` inside `interpolation.py`. Just as `interpolate_bads_meeg()` works inside `interpolation.py`.

larsoner commented:
> Shall I add tests? or is the example script enough?

I think ideally here you would modify some example, and some tests. For a simple test, if you take Neuromag data, transform it to CTF, and back to the original Neuromag info / dev_head_t, are the start and end data very highly correlated?

I would also check that the montages loaded for Neuromag match the `.info` from some test file up to numerical precision. Same thing for any other systems you add.

For failed tests you should be able to click on the name of any failed CI. So for example "ubuntu-latest pip" which I know runs almost all the tests, you can see a few failures:

https://github.com/mne-tools/mne-python/actions/runs/18536524047/job/52833052296?pr=13196#step:19:6190

contsili commented:
> For failed tests you should be able to click on the name of any failed CI. So for example "ubuntu-latest pip" which I know runs almost all the tests, you can see a few failures:
> 
> https://github.com/mne-tools/mne-python/actions/runs/18536524047/job/52833052296?pr=13196#step:19:6190

The tests fail because the neuromag and ctf montages do not have the dig attribute:

`FAILED mne/channels/tests/test_standard_montage.py::test_standard_montages_have_fids[ctf275] - AttributeError: 'CustomMontage' object has no attribute 'dig'`

I think I need to make some changes to `read_meg_montage `to fix this. Would it make sense to add `read_meg_montage `directly in `channels.py` as a helper function, since MEG sensors are not real montages and should not be handled as such within `montage.py`? This seems to me the most straightforward solution, but I’d like your advice since you know the MNE architecture better.

larsoner commented on mne/channels/montage.py:
```diff
@@ -1681,6 +1683,99 @@ def read_custom_montage(
+
+    # Read sensor definitions
+    # TODO: Refactor not to use Pandas (should be able to parse without it)
+    df = pd.read_csv(csv_file)
```
`pandas` is an optional dependency of MNE and not everyone will have it. Is it easy enough to just parse the CSV directly / manually? Usually we try to do that if it's not too much work

larsoner commented:
@contsili do you have time to work on this? If it would help I could push some commits as well

larsoner commented on mne/channels/interpolation.py:
```diff
@@ -415,3 +415,207 @@ def _interpolate_bads_seeg(
+    inst : instance of Raw, Epochs, or Evoked
+        A new instance with interpolated data.
+    """
+    from .._fiff.meas_info import create_info
```
CIs complain that some imports are nested that should not be etc.

larsoner commented on mne/channels/data/canonical_meg/ctf151.csv:
```diff
@@ -0,0 +1,152 @@
+name,coil_type,x,y,z,ex_x,ex_y,ex_z,ey_x,ey_y,ey_z,ez_x,ez_y,ez_z
```
@contsili I moved these to their own directory

I also renamed `read_meg_montage` to `read_meg_canonical_info` since it returns an `Info` not a montage

chsquare commented:
Hi all! Thanks for implementing this transformation! I just noticed that interpolate_to() resets sampling frequency, filter info, and other information which should instead be maintained across the interpolation step. I will open pull request containing a fix.
